erdos/src/dataflow/graph/graph.rs

```diff
@@ -0,0 +1,342 @@
+use std::sync::{Arc, Mutex};
+
+use serde::Deserialize;
+
+use crate::{
+    dataflow::{
+        operator::{
+            OneInOneOut, OneInTwoOut, ParallelOneInOneOut, ParallelOneInTwoOut, ParallelSink,
+            ParallelTwoInOneOut, Sink, Source, TwoInOneOut,
+        },
+        stream::{ExtractStream, IngestStream, OperatorStream},
+        AppendableState, Data, LoopStream, State, Stream,
+    },
+    OperatorConfig,
+};
+
+use super::{InternalGraph, JobGraph};
+
+pub struct Graph {
+    internal_graph: Arc<Mutex<InternalGraph>>,
+}
+
+impl Default for Graph {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Graph {
+    pub fn new() -> Self {
+        Self {
+            internal_graph: Arc::new(Mutex::new(InternalGraph::new())),
+        }
+    }
+
+    pub fn get_ingest_stream<D>(&self, name: &str) -> IngestStream<D>
+    where
+        for<'a> D: Data + Deserialize<'a>,
+    {
+        let ingest_stream = IngestStream::new(name, Arc::clone(&self.internal_graph));
+        self.internal_graph
+            .lock()
+            .unwrap()
+            .add_ingest_stream(&ingest_stream);
+
+        ingest_stream
+    }
+
+    pub fn get_extract_stream<D>(&self, name: &str, stream: &OperatorStream<D>) -> ExtractStream<D>
```

Comment:
API-wise, it might make more sense to add a `to_extract_stream()` function to the `OperatorStream` struct, and make this function `pub(crate)`.

Also, I don't think it makes sense to give the extract stream a separate name, as it's like a `ReadStream` that happens to live in the driver, rather than a whole different stream that originates from an operator.

Reply:
Agree with the first, we should probably remove `get_extract_stream`, and rename it to `extract`, which then also allows us to provide utility functions like `extract_to_file` to log stream contents to the disk etc.